python/ray/train/v2/_internal/execution/train_fn_utils.py

```diff
@@ -68,6 +77,51 @@ def get_dataset_shard(self, dataset_name: str) -> DataIterator:
     def get_context(self) -> ExternalTrainContext:
         return ExternalTrainContext()
 
+    def barrier(self) -> None:
+        """Create a barrier across all workers.
+
+        All workers must call this method before the training function can continue.
+
+        This method is used by the public API function :func:`ray.train.collective.barrier`.
+        Users should typically call ``ray.train.collective.barrier()`` instead of calling this method directly.
+        """
+        train_context = get_internal_train_context()
+        sync_actor = train_context.get_synchronization_actor()
+        return ray.get(
+            sync_actor.broadcast_from_rank_zero.remote(
+                world_rank=train_context.get_world_rank(),
+                world_size=train_context.get_world_size(),
+                data=None,
+                caller_method_name="ray.train.collective.barrier",
+            )
```

Comment:
I don't want the collective implementations in train_fn_utils -- I still want those in in the `collective/` folder.

We can do this:
```
python/ray/train/collective/
    api.py
    collectives_impl.py    
```

api.py has the public-facing method w/ the docstring and api stability label
which points to train_fn_utils which is used for group patching
which defaults to pointing to the collective implementations